### ISSUE TITLE
TextFileReader needed?

### ISSUE THREAD

**Issue by ludwiglierhammer** [glamod/cdm_reader_mapper#8]
### Current state

Optionally, you can select a chunksize when reading the data. If this is the case, the data is read as a ``TextFileReader`` object for iterations; see [pandas.read_csv](https://pandas.pydata.org/pandas-docs/version/1.3/reference/api/pandas.read_csv.html). This complicates the code because manipulating ``TextFileReader`` objects causes iterations and high I/O loads. In earlier ``pandas`` versions, reading chunked data freed up memory and made the procedure faster.

### Question

Is it necessary to make use of chunked ``TextFileReader`` objects are can newer ``pandas`` versions can handle this issue?

### Possible Solution

Get rid of ``TextFileReader`` objects.

### Advantage

Simplify the code.

**Comment by ludwiglierhammer** [comment]
Hi @jtsiddons, do you set chunksizes when using the ``mdf_reader``? Because newer ``pandas`` versions can process large inputs, I think we do not need this anymore.

**Comment by jtsiddons** [comment]
Hi @ludwiglierhammer, we have on ocassion used the chunk-size component - mainly for the larger ICOADS NRT files, or ICOADS total files after 2000 or so. The reader for these files seem to use _a lot_ of memory, I'm not sure the chunksize parameter makes much difference.

I _think_ the memory usage problems occur after the initial read - do you know if the chunks are combined after the initial read?

Can we remove the python 3.11.9 issue from the README? I've not experienced any issues on 3.11.9. Is 3.11.9 the version in the ci-tests?

**Comment by ludwiglierhammer** [comment]
@jtsiddons: The chunks are not combined after the initial read. When manipulating the ``TextFileReader`` object it is opened chunk by chunk. For each chunk the manipulation will be done. Then the manipulated chunk is saved in a StringIO buffer and written to disk.
Afterwards the file on disk is read in again as a ``TextFileReader`` object. Is yields high I/O loads.

E.g. see the [difference](https://github.com/glamod/cdm_reader_mapper/blob/main/cdm_reader_mapper/mdf_reader/utils/auxiliary.py#L511-L579) between ``pd.DataFrame`` and ``TextFileReader`` objects.

Can you please test your larger ICOADS files without the chunk-size component? Do you see any differences?

**Comment by ludwiglierhammer** [comment]
@jtsiddons: The python 3.11. test succesfully runs with python 3.11.9. I'll remove the python 3.11.9 issue from the README. Thanks for the hint.

**Comment by jtsiddons** [comment]
@ludwiglierhammer: Thanks for the info. I'll run a larger file on one of our servers on Monday morning.

**Comment by jtsiddons** [comment]
@ludwiglierhammer Do you want me to test on main or v0.3.0? I can look at running `mdf_reader.read` through the `memory-profiler` package too

e: branch name

**Comment by ludwiglierhammer** [comment]
@jtsiddons:You can run the tests on your fork. It does not matter which branch you actually use. No changes have been made to the corresponding code snippets since v3.0.0. 

Do you mean this package: https://pypi.org/project/memory-profiler/
Unfortunately, this is no longer maintained. I do not know this package. Is it trustworthy?

**Comment by jtsiddons** [comment]
@ludwiglierhammer - good spot - I hadn't noticed it was no-longer maintained. I've not had any issues with it in the past. I'll leave the profiler for now.

**Comment by jtsiddons** [comment]
@ludwiglierhammer: Update from run with `chunksize=None`. I have cancelled the job it was close to memory limits (~200G). It did not get past "Getting data string from source stage". I did select one of the largest ICOADS NRT Total files.

Manual reporting, data from `top`.
Python : 3.11.3
OS: Rocky 9

```py
in_file = "./IMMA1_R3.0.2T_2023-12"
out_dir = "./mem_test_out"

chunksize = None
mdf_reader.read(in_file, data_model="imma1", out_path=out_dir, chunksize=chunksize)
```

File: `IMMA1_R3.0.2T_2023-12`
Size on disk : 3.0G (3_117_986_480)
Lines in file: 6_195_778

`chunksize=None` : 
| Stage                             | VIRT   | RES    | TIME+ |
| ---                               | ---:   | ---:   | ---:  |
| "Getting data string from source" | 65.2g  | 64.0g  | 30:00 |
| ^^^                               | 95.9g  | 94.4g  | 45:04 |
| ^^^                               | 127.0g | 125.1g | 60:01 |
| ^^^                               | 157.7g | 155.4g | 75:00 |
| ^^^                               | 180.1g | 177.7g | 90:02 |

``` 
File "CDM_READER_MAPPER/./test_mem_use.py", line 17, in <module>
    main()
  File "CDM_READER_MAPPER/./test_mem_use.py", line 12, in main
    mdf_reader.read(in_file, data_model="imma1", out_path=out_dir, chunksize=chunksize)
  File "CDM_READER_MAPPER/cdm_reader_mapper/mdf_reader/read.py", line 321, in read
    return MDFFileReader(
           ^^^^^^^^^^^^^^
  File "CDM_READER_MAPPER/cdm_reader_mapper/mdf_reader/read.py", line 241, in read
    self.data, self.isna = self._open_data(
                           ^^^^^^^^^^^^^^^^
  File "CDM_READER_MAPPER/cdm_reader_mapper/mdf_reader/utils/auxiliary.py", line 512, in _open_data
    df, self.missings = self._read_sections(
                        ^^^^^^^^^^^^^^^^^^^^
  File "CDM_READER_MAPPER/cdm_reader_mapper/mdf_reader/utils/auxiliary.py", line 476, in _read_sections
    df = TextParser.apply(
         ^^^^^^^^^^^^^^^^^
  File "lib/python3.11/site-packages/pandas/core/frame.py", line 9423, in apply
    return op.apply().__finalize__(self, method="apply")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "lib/python3.11/site-packages/pandas/core/generic.py", line 5931, in __finalize__
    @final

KeyboardInterrupt
```

I will run with `chunksize` set this afternoon.

e: complete incomplete sentence

**Comment by jtsiddons** [comment]
Results with `chunksize`. The reading step uses much less memory, so obviously this does have an effect. However, even with `chunksize` I would not be able to run this on a compute/data centre server such as JASMIN for example, where memory limits tend to apply.

`chunksize=100_000` :
NOTE - Got to > 100g for a few seconds at 117:00 (end of read?)
NOTE - Got to > 150g for a few seconds at 162:00 (end of validation?)
| Stage (approximate)               | VIRT   | RES    | TIME+  |
| ---                               | ---:   | ---:   | ---:   |
| "Getting data string from source" | 9.9g   | 9.3g   | 29:43  |
| ^^^                               | 12.9g  | 12.3g  | 45:00  |
| ^^^                               | 16.0g  | 15.4g  | 61:30  |
| ^^^                               | 18.7g  | 18.1g  | 75:00  |
| ^^^                               | 21.6g  | 21.0g  | 90:00  |
| ^^^                               | 25.5g  | 24.8g  | 108:38 |
| ^^^ - "Convert/Decode"?           | 57.0g  | 56.4g  | 116:54 |
| ^^^                               | 71.1g  | 70.5g  | 117:57 |
| "Validate Entries"                | 108.6g | 108.0g | 141:57 |
| "Validate Entries"                | 113.0g | 112.4g | 159:43 |
| "Writing Data Files"              | 144.3g | 143.7g | 162:54 |

Total elapsed ~ 175 mins

I may re-run the test at some-point for a (large) subset of the data (say 2 million rows), to see how the post-read stages look memory-wise.

**Comment by jtsiddons** [comment]
Final output size on disk:

```
 54K Jun 24 15:41 atts.json
4.0G Jun 24 15:41 data.csv
8.9G Jun 24 15:41 mask.csv
```

**Comment by ludwiglierhammer** [comment]
@jtsiddons: Thanks for your tests. We have to think how we deal with such large files. Maybe we should use [dask.DataFrames](https://docs.dask.org/en/latest/dataframe.html) instead of pandas DataFrames. One can simply read csv tables in a dask DataFrame using [dask.dataframe.read_csv](https://docs.dask.org/en/stable/generated/dask.dataframe.read_csv.html). I think this could speed up processing large files and reduce the memory load.

**Comment by jtsiddons** [comment]
@ludwiglierhammer: I've performed a few more experiments reading large files. I think I may have identified one of the causes of high memory usage during reading.

https://github.com/glamod/cdm_reader_mapper/blob/main/cdm_reader_mapper/mdf_reader/utils/auxiliary.py#L501-L503

When splitting the lines into sections/elements, a `Configurator` class is initialised in a `DataFrame.apply` call (which applies `Configurator.read()` to each row in the DataFrame). I think memory is allocated for each `Configurator` class instance (which includes the schema). A `Configurator` class is initialised for _each_ line in the data.

In my testing, the memory is eventually returned, but doesn't occur until after the `apply` method is finished (at the earliest).

**Comment by ludwiglierhammer** [comment]
@jtsiddons, the reason why a ``Configurator`` class is initialised for **each** row is that it could be that the structure of each row is different or that we have a combination of fixed-width sections and comma-seperated sections. Perhaps we should abandon this theoretical consideration. While rewriting the old code, I already stumbled across these lines. Then I decided to adopt the old line-by-line reading. 

Could you try out a new PR without this line-by-line reading? And do you know of any cases where line-by-line reading is necessary? I am not very familiar with the ICOADS format.

**Comment by ludwiglierhammer** [comment]
And another probelm is that many ``c99 attachments`` are comma-seperated lists while all the other sections are mostly fixed-width lists. Perhaps we should read the data in two steps.

**Comment by ludwiglierhammer** [comment]
For instance, in [icoads_r300_d702](https://github.com/glamod/cdm_reader_mapper/blob/main/cdm_reader_mapper/mdf_reader/schemas/icoads/r300/d702/icoads_r300_d702.json) comma-separated lists and fixed-width list are mixed in the c99 attachment.

**Comment by ludwiglierhammer** [comment]
For the sake of simplicity, bycomma-separated list I mean any list that is separated by a delimiter.

**Comment by jtsiddons** [comment]
> Could you try out a new PR without this line-by-line reading? And do you know of any cases where line-by-line reading is necessary? I am not very familiar with the ICOADS format.

Sure. I'll look into this over the course of the week.

**Comment by ludwiglierhammer** [comment]
In PR #23 I've already deleted code lines containing TextFileReader if clauses. This PR only works with pandas DataFrame. I haven't merged it because I don't know if it works for large data. Maybe you can have a look on it or create a PR to this PR. You can clone [my fork](https://github.com/ludwiglierhammer/cdm_reader_mapper). Perhaps it makes sense to work together on this PR.

**Comment by ludwiglierhammer** [comment]
I want to replacre pandas DataFrames with [dask DataFrames](https://docs.dask.org/en/stable/dataframe.html) in this PR. I should enhance the perfomance.

**Comment by jtsiddons** [comment]
> In PR #23 I've already deleted code lines containing TextFileReader if clauses. This PR only works with pandas DataFrame. I haven't merged it because I don't know if it works for large data. Maybe you can have a look on it or create a PR to this PR. You can clone [my fork](https://github.com/ludwiglierhammer/cdm_reader_mapper). Perhaps it makes sense to work together on this PR.

I can probably test a large file on our system with your PR, and again with Dask.

I am currently running other processing at the moment, so it won't be until that is freed up.

**Comment by ludwiglierhammer** [comment]
That would be a great help. Don't hurry, we don't perform open heart surgery.
Thanks a lot!

**Comment by ludwiglierhammer** [comment]
FYI: At end of this week or the beginning of the next week, I'll make a new ``cdm_reader_mapper`` release, publish it on pypi and assign a zenodo DOI.

**Comment by jtsiddons** [comment]
Will that be v0.4.0? or v1.0.0?

**Comment by ludwiglierhammer** [comment]
Interesting that you ask. I am thinking of v1.0.0 because with this version we want to calculate our new GLAMOD release. Do you have any preferences?

**Comment by jtsiddons** [comment]
That's fine with me.

Do you want to have the `TextFileReader` / `Dask` / large file changes implemented in this release, or are you considering the current state to be a _release candidate_ with no new features/significant changes to be implemented?

**Comment by ludwiglierhammer** [comment]
I am considering the current state to be the new release. We can solve this issue after our GLAMOD release. I am still testing the cuurent version. Hopefully, everything works.

**Comment by jtsiddons** [comment]
> @jtsiddons, the reason why a `Configurator` class is initialised for **each** row is that it could be that the structure of each row is different or that we have a combination of fixed-width sections and comma-seperated sections. Perhaps we should abandon this theoretical consideration. While rewriting the old code, I already stumbled across these lines. Then I decided to adopt the old line-by-line reading.
> 
> Could you try out a new PR without this line-by-line reading? And do you know of any cases where line-by-line reading is necessary? I am not very familiar with the ICOADS format.

Since an imma file will likely contiain lines with different subsets of the attachments/sections, one would need a way of identifying which attachments are in any given line without reading line-by-line. I am not sure how one would approach that. 

I am working on a branch on my fork where I have re-factored the `Configurator` class to take the entire pandas frame as input. The `open_pandas` method is then a wrapper for a `_read_line` method. This means that the `Configurator` class is initialised only once.

Currently, it is ugly but it passes the unit tests when working in my shell (I've been cancelling the runs on github for now as it's very much a WIP). I need to run this with a larger file on our servers to see if I have improved the memory usage of the reading part (I've not looked at the decoding, conversion, or validation steps).

If it improves memory/performance & I can clean it up _significantly_ I can create a PR.

Current state: https://github.com/jtsiddons/cdm_reader_mapper/compare/main...jtsiddons:cdm_reader_mapper:configurator_refactor

**Comment by ludwiglierhammer** [comment]
That looks good to my at the first glance.  I am quiet busy at the moment cause we have to deliver our new GLAMOD release. I'll come back to you asap.

**Comment by ludwiglierhammer** [comment]
hi @jtsiddons, does your ``configurator_refactor`` improve memory and performance? If so please create a `PR`. I think it is easier to discuss in an open `PR`. We've finished our GLAMOD release and I have time for coding :blush:.

**Comment by jtsiddons** [comment]
@ludwiglierhammer: Sorry, I haven't had chance to test on a larger file yet. I may not get chance until early January I'm afraid

**Comment by ludwiglierhammer** [comment]
> @ludwiglierhammer: Sorry, I haven't had chance to test on a larger file yet. I may not get chance until early January I'm afraid

No probelm ☺️

**Comment by jtsiddons** [comment]
Hi @ludwiglierhammer 

I've finally had chance to test the refactor of the `Configurator` class on some larger files. To summarise, the change did not improve memory usage, there was a minor speed improvement (only having to initialise a single `Configurator` instance).

Probably not worth continuing with my refactor of the `Configurator`.

## Results

Note that the below is with

```python
read(file, imodel="icoads", convert=False, decode=False, validator=False)
```

so that I am only testing the initial read stage.

### Current `Configurator` (from main branch)

| Number of lines | Peak Memory (GB) |
| ---:            | ---              |
| 100000          | 4.9              |
| 250000          | 11.8             |
| 500000          | 23.2             |
| 1000000         | 46.8             |

### Refactored `Configurator`

| Number of lines | Peak Memory (GB) |
| ---:            | ---              |
| 100000          | 4.9              |
| 250000          | 11.8             |
| 500000          | 23.3             |
| 1000000         | 46.5             |

This was not a true memory test, as I just logged peak memory from the process on `top`. So the differences will be a consequence of the sampling frequency by `top` (this was 1 second).

### Time

To process 1 million rows:

* Current `Configurator`: 19 mins
* Refactored `Configurator`: 17 mins

I did not note down times for the smaller files, although all 4 cases were slower with the current `Configurator`.

**Comment by ludwiglierhammer** [comment]
@jtsiddons: Many thanks for your tests. Could you please run the tests with the parameter ``chunksize`` to test the perfomance difference bwtween with and without chunked ``TextFileReader`` objects? In the [glamod-marine-processing scripts](https://github.com/glamod/glamod-marine-processing/blob/main/glamod_marine_processing/obs_suite/scripts/_utilities.py#L30-L34) we use a chunksize of ``200000``. Honestly, I am not sure if we need the chunking while reading large data at all.

The next step would be to test the entire workflow: reading, mapping, writing:

```
from cdm_reader_mapper import read_mdf
source = <your-large-file>
imodel="icoads"

db = read_mdf(source, imodel=imodel)
db.map_model()
db.write_tables(suffix=imodel)
```

**Comment by ludwiglierhammer** [comment]
In the meanwhile, I try to replace ``pd.DataFrame``s with ``dask.Dataframe``.

**Comment by jtsiddons** [comment]
> [@jtsiddons](https://github.com/jtsiddons): Many thanks for your tests. Could you please run the tests with the parameter `chunksize` to test the perfomance difference bwtween with and without chunked `TextFileReader` objects? In the [glamod-marine-processing scripts](https://github.com/glamod/glamod-marine-processing/blob/main/glamod_marine_processing/obs_suite/scripts/_utilities.py#L30-L34) we use a chunksize of `200000`. Honestly, I am not sure if we need the chunking while reading large data at all.

Sure thing. I'll try to run some tests this afternoon if there is opportunity on our servers. If not it'll be Thursday as I'm at a meeting Mon/Tues and on leave on Weds.

**Comment by ludwiglierhammer** [comment]
This is [my test branch](https://github.com/ludwiglierhammer/cdm_reader_mapper/tree/dask_dataframe) for using ``dask.DataFrame`` instead of ``pandas.DataFrame``.

**Comment by jtsiddons** [comment]
@ludwiglierhammer: I've run some tests for the reader for different `chunksize` settings. Note that I am only doing the read stage, I set `convert`, `decode`, and `validate` arguments to `False`.

## Chunk Testing

Using **main** branch (commit: e97d666)

Using `time --verbose` to track.

- Peak Memory Use is "maximum resident set size"
- Time Elapsed is "Elapsed (wall clock) time"

| Num Rows  | Chunksize | Peak Memory Use (read) (kB) | Time Elapsed |
| ---:      | ---:      | ---:                        | ---          |
| 100_000   | N/A       | 5_418_592                   | 2:07.83      |
| 100_000   | 25_000    | 3_418_860                   | 2:22.84      |
| 250_000   | N/A       | 12_754_380                  | 5:21.68      |
| 250_000   | 50_000    | 6_773_956                   | 5:57.19      |
| 500_000   | N/A       | 25_149_700                  | 10:43.52     |
| 500_000   | 100_000   | 13_327_444                  | 11:54.01     |
| 1_000_000 | N/A       | 49_991_300                  | 21:36.40     |
| 1_000_000 | 200_000   | 26_423_984                  | 24:22.02     |
| 1_000_000 | 50_000    | 19_007_680                  | 23:35.07     |

I additionally ran the **configurator_refactor** branch for 1 million rows:

| Num Rows  | Chunksize | Peak Memory Use (read) (kB) | Time Elapsed |
| ---:      | ---:      | ---:                        | ---          |
| 1_000_000 | N/A       | 49_989_828                  | 19:46.78     |

## Test Code

```python
#!/usr/bin/env python

from cdm_reader_mapper import read_mdf, DataBundle
import os
import pandas as pd


def main() -> None:
    n: int = 1_000_000
    chunksize: int | None = None

    base_path: str = os.path.dirname(__file__)
    file: str = os.path.join(base_path, "scratch", f"test_file_{n}")

    if not os.path.isfile(file):
        raise FileNotFoundError(f"Cannot find file with {n} lines, {file}")

    out: DataBundle = read_mdf(
        file,
        imodel="icoads",
        convert=False,
        decode=False,
        validate=False,
        chunksize=chunksize,
    )

    return None


if __name__ == "__main__":
    main()
```

**Comment by ludwiglierhammer** [comment]
@jtsiddons: Thank you for testing different ``chunksize`` settings. That was very helpful. Since we can reduce the memory peak by more than half, I think we need these chunked ``TextFileReader`` objects at the moment. I'll get back to you  as soon as I have some new information about the ``dask.DataFrame`` branch.

**Comment by ludwiglierhammer** [comment]
@jtsiddons: Could you please create a ``PR`` from your [``configurator_refactor`` branch](https://github.com/jtsiddons/cdm_reader_mapper/tree/configurator_refactor) to my [``no_textfilereader`` branch](https://github.com/ludwiglierhammer/cdm_reader_mapper/tree/no_textfilereader). I like the way you simplified the reading.

**Comment by ludwiglierhammer** [comment]
@jtsiddons: Or may I copy your changes directly into my branch myself?

**Comment by ludwiglierhammer** [comment]
@jtsiddons: if it is ok for you, I would merge your changes (I copied them into a [new PR ](https://github.com/ludwiglierhammer/cdm_reader_mapper/pull/2)) into my ``no_textfilereader`` branch. Your changes would probably simplify the implementation of ``dask.DataFrames``.

**Comment by jtsiddons** [comment]
Hi @ludwiglierhammer, apologies, I have been at a meeting this week and didn't have my laptop with me.

Thank you for duplicating the refactor. I'll still update my branch to align with the changes to main branch (splitting auxilary.py into 2 files) in case it is still needed.

I plan to run similar tests as before but run the decoding, conversion, and validation steps. I also hope to run the reader through a memory profiler (such as memray: https://github.com/bloomberg/memray) to get a more in-depth breakdown of calls and memory usage.

**Comment by ludwiglierhammer** [comment]
Hi @jtsiddons, no worries 😊. I fight enough tough battles with ``dask.DataFrames``.

**Comment by ludwiglierhammer** [comment]
I think I'll give [polars](https://pola.rs/). Ever heard of it?

**Comment by jtsiddons** [comment]
I prefer polars to pandas!

**Comment by jtsiddons** [comment]
I've updated my configurator branch, I think I've merged main in correctly, it pretty much required a duplication of the changes. 

Do you still want me to do a PR with it?

> I think I'll give [polars](https://pola.rs/). Ever heard of it?

I use polars almost exclusively for DataFrame work. It is quite different to pandas and could involve some significant changes.

I did start writing a (specifically imma) reader in polars at one point but never completed it and dropped the project.

I'm happy to discuss polars at some point if you want.

**Comment by jtsiddons** [comment]
I have run the basic memory tests with `decode=True`, `convert=True`, and `validate=True`. Not too much difference against the previous tests.

| Num Rows  | Chunksize | Peak Memory Use (read) (kB) | Time Elapsed |
| ---:      | ---:      | ---:                        | ---          |
| 100_000   | N/A       | 5_419_892                   | 2:39.49      |
| 100_000   | 25_000    | 3_437_688                   | 3:03.19      |
| 1_000_000 | N/A       | 49_998_416                  | 25:09.30     |
| 1_000_000 | 200_000   | 28_568_316                  | 31:28.79     |
| 1_000_000 | 50_000    | 25_798_004                  | 32:12.16     |

**Comment by ludwiglierhammer** [comment]
> Do you still want me to do a PR with it?

Many thanks for updating your branch. Yes please, but do a PR for th ``main`` branch not my ``no_textfilereader`` branch.

**Comment by ludwiglierhammer** [comment]
> I'm happy to discuss polars at some point if you want.

I tried to implement polars yesterday withour any success 😞. As you said it is quite different to pandas and the implementation needs some more effort. Maybe we should discuss how to imporve the performance in general. What about end of next week?

**Comment by jtsiddons** [comment]
> > Do you still want me to do a PR with it?
> 
> Many thanks for updating your branch. Yes please, but do a PR for th `main` branch not my `no_textfilereader` branch.

See #208

**Comment by ludwiglierhammer** [comment]
> I plan to run similar tests as before but run the decoding, conversion, and validation steps. I also hope to run the reader through a memory profiler (such as memray: https://github.com/bloomberg/memray) to get a more in-depth breakdown of calls and memory usage.

This would be pretty good. I think the main bottelneck is reading the files. 

I have a new [PR](https://github.com/glamod/cdm_reader_mapper/pull/207) where I tried an in-situ conversion, decoding and validation of the data while reading. Until now several bolean masks of the data are created while reading. Those mask are then used in  the validation scheme. Even if ``validation=False``. I tried to get rid of those masks.

You may want to test whether this leads to any improvements regarding the performance/memory usage.
But this is the first version. I have to to some more tidy uos/improvements if it really helps to improve the perfomance. 

Note: It only run with ``validation=True`` at the moment.

**Comment by jtsiddons** [comment]
> > I'm happy to discuss polars at some point if you want.
> 
> I tried to implement polars yesterday withour any success 😞. As you said it is quite different to pandas and the implementation needs some more effort. Maybe we should discuss how to imporve the performance in general. What about end of next week?

End of the next week is good for me. Usually I'm the only person in my office on a Friday so that would be most convenient. Drop me an email with some times that work for you.

**Comment by jtsiddons** [comment]
> > I plan to run similar tests as before but run the decoding, conversion, and validation steps. I also hope to run the reader through a memory profiler (such as memray: https://github.com/bloomberg/memray) to get a more in-depth breakdown of calls and memory usage.
> 
> This would be pretty good. I think the main bottelneck is reading the files.
> 
> I have a new [PR](https://github.com/glamod/cdm_reader_mapper/pull/207) where I tried an in-situ conversion, decoding and validation of the data while reading. Until now several bolean masks of the data are created while reading. Those mask are then used in the validation scheme. Even if `validation=False`. I tried to get rid of those masks.
> 
> You may want to test whether this leads to any improvements regarding the performance/memory usage. But this is the first version. I have to to some more tidy uos/improvements if it really helps to improve the perfomance.
> 
> Note: It only run with `validation=True` at the moment.

I'll look at testing this next week

**Comment by jtsiddons** [comment]
@ludwiglierhammer: I am currently experimenting with a `Configurator.open_polars` method. I'll update you if I get anywhere with it.

**Comment by jtsiddons** [comment]
Working on a branch on my fork for `polars` - https://github.com/jtsiddons/cdm_reader_mapper/tree/polarising

It's partially working so far, I've only worked on the section reading. Still a fair bit to do and it does not pass unit tests at present. If I can get it working properly I'll do a PR and we can look at refactoring the decoding/conversion/validation steps.

**Comment by ludwiglierhammer** [comment]
That's pretty good! You could start a PR now, That make the discussion easier and this issue not too cluttered. You can also create a [draft PR](https://github.blog/news-insights/product-news/introducing-draft-pull-requests/) as a first attempt. If it works, you can convert it into a "normal" PR.

**Comment by jtsiddons** [comment]
Will do.

I've just run some performance tests on the branch as it currently is, it is faster and memory use is a little lower. There is a caveat that virtual memory is significantly higher. There is a noticeable ramp up in memory usage _after_ `Configurator.open_polars` (and also `Configurator.open_pandas`), but before any decoding/conversion/validation.

| Num Rows  | Chunksize | Peak Memory Use (read) (kB) | Time Elapsed |
| ---:      | ---:      | ---:                        | ---          |
| 100_000   | N/A       | 4_843_244                   | 1:52.49      |
| 500_000   | N/A       | 22_702_756                  | 9:29.20      |

The above tests were carried out with `decode=False`, `convert=False`, and `validate=False`. Chunking does work in the current state.

**Comment by jtsiddons** [comment]
I did a quick test, raising an error before `set_missing` is called (i.e. after `Configurator` finishes). 

| Num Rows | Stage              | Peak Memory Use (read) (kB) |
| ---:     | ---:               | ---:                        |
| 100_000  | After Set Missing  | 5_418_976                   |
| 100_000  | After Configurator | 3_665_220                   |

The above is on **main** branch.

**Comment by ludwiglierhammer** [comment]
``Configurator.open_pandas`` creates a DataFrame with the data plus a new column called ``missing_values``. This column contains a list of column names per row that are missing or not representented in the input data. The line 154 in ``filereader.py`` [missing_values = set_missing_values(pd.DataFrame(missing_values_), df)](https://github.com/glamod/cdm_reader_mapper/blob/main/cdm_reader_mapper/mdf_reader/utils/utilities.py#L77-L88) converts this series of lists into a boolean DataFrame with the shape of data. I think this line leads to the noticeable ramp up in memory usage you mentioned. This DataFrame is later used by the validation scheme. The creation of this second DataFrame is independent of the parameters ``decode``, ``convert`` and ``validate``. In my [performance PR](https://github.com/glamod/cdm_reader_mapper/pull/207) I try to avoid creating a second DataFrame, by doing the decoding, conversion and validation directly while reading the data.

**Comment by jtsiddons** [comment]
> `Configurator.open_pandas` creates a DataFrame with the data plus a new column called `missing_values`. This column contains a list of column names per row that are missing or not representented in the input data. The line 154 in `filereader.py` [missing_values = set_missing_values(pd.DataFrame(missing_values_), df)](https://github.com/glamod/cdm_reader_mapper/blob/main/cdm_reader_mapper/mdf_reader/utils/utilities.py#L77-L88) converts this series of lists into a boolean DataFrame with the shape of data. I think this line leads to the noticeable ramp up in memory usage you mentioned. This DataFrame is later used by the validation scheme. The creation of this second DataFrame is independent of the parameters `decode`, `convert` and `validate`. In my [performance PR](https://github.com/glamod/cdm_reader_mapper/pull/207) I try to avoid creating a second DataFrame, by doing the decoding, conversion and validation directly while reading the data.

This may be an easier propsition in polars as we can work column-wise rather than iterating over the rows.

**Comment by ludwiglierhammer** [comment]
> I did a quick test, raising an error before `set_missing` is called (i.e. after `Configurator` finishes).
> 
> Num Rows	Stage	Peak Memory Use (read) (kB)
> 100_000	After Set Missing	5_418_976
> 100_000	After Configurator	3_665_220
> The above is on **main** branch.

👍 so the `set_missing` is one of our little troublemakers. What about my PR? Any improvements?

**Comment by jtsiddons** [comment]
Just to check: I need to run with `decode=True`, `convert=True`, and `validate=True` or just `validate=True`?

**Comment by ludwiglierhammer** [comment]
In a first test I would set all three parameters to `False`.
Maybe you could do a second one with all three parameters set to `True`.

**Comment by jtsiddons** [comment]
@ludwiglierhammer: Failed to complete: 

```
Traceback (most recent call last):
  File "<<TRIMMED_PATH>>/git/cdm_ll/./test_cdm_mem.py", line 32, in <module>
    main()
  File "<<TRIMMED_PATH>>/git/cdm_ll/./test_cdm_mem.py", line 19, in main
    out: DataBundle = read_mdf(
                      ^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/cdm_reader_mapper/mdf_reader/read.py", line 187, in read_mdf
    return MDFFileReader(
           ^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/cdm_reader_mapper/mdf_reader/read.py", line 97, in read
    self.data, self.mask = self.open_data(
                           ^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/cdm_reader_mapper/mdf_reader/utils/filereader.py", line 176, in open_data
    df, mask = self._read_sections(
               ^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/cdm_reader_mapper/mdf_reader/utils/filereader.py", line 133, in _read_sections
    df = df.astype(configurations["dtype"])
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/generic.py", line 6620, in astype
    res_col = col.astype(dtype=cdt, copy=copy, errors=errors)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/generic.py", line 6643, in astype
    new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/internals/managers.py", line 430, in astype
    return self.apply(
           ^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/internals/managers.py", line 363, in apply
    applied = getattr(b, f)(**kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/internals/blocks.py", line 758, in astype
    new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/dtypes/astype.py", line 237, in astype_array_safe
    new_values = astype_array(values, dtype, copy=copy)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/dtypes/astype.py", line 182, in astype_array
    values = _astype_nansafe(values, dtype, copy=copy)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/dtypes/astype.py", line 80, in _astype_nansafe
    return dtype.construct_array_type()._from_sequence(arr, dtype=dtype, copy=copy)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/arrays/masked.py", line 152, in _from_sequence
    values, mask = cls._coerce_to_array(scalars, dtype=dtype, copy=copy)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/arrays/numeric.py", line 272, in _coerce_to_array
    values, mask, _, _ = _coerce_to_data_and_mask(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<<TRIMMED_PATH>>/git/cdm_ll/venv/lib/python3.11/site-packages/pandas/core/arrays/numeric.py", line 227, in _coerce_to_data_and_mask
    values = values.astype(dtype, copy=copy)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: invalid literal for int() with base 10: 'A': Error while type casting for column '('c5', 'IC2')'
```

Also, setting all arguments to True appears to repeat the logging warnings (possibly every row? I cancelled that test)

**Comment by ludwiglierhammer** [comment]
Sorry, my bad. Do you know the row wehrer you get this error message? Perhaps you can send it to me for debugging.

Addiotionally, I should delete the logging warning.

**Comment by ludwiglierhammer** [comment]
Do you get the error as well if all arguments are set to `True`?

**Comment by jtsiddons** [comment]
> Sorry, my bad. Do you know the row wehrer you get this error message? Perhaps you can send it to me for debugging.
> 
> Addiotionally, I should delete the logging warning.

I've invited you to a repo that includes my test script and some of the memory testing files (the 10_000 and 100_000 row versions)

**Comment by jtsiddons** [comment]
> Do you get the error as well if all arguments are set to `True`?

It does not raise an error if all arguments are set to `True`. Although the logging **really** slows it down. `10_000` rows took 19:10.06 to complete!

**Comment by ludwiglierhammer** [comment]
> > Do you get the error as well if all arguments are set to `True`?
> 
> It does not raise an error if all arguments are set to `True`. Although the logging **really** slows it down. `10_000` rows took 19:10.06 to complete!

Issues should be fixed.

**Comment by jtsiddons** [comment]
> > > Do you get the error as well if all arguments are set to `True`?
> > 
> > 
> > It does not raise an error if all arguments are set to `True`. Although the logging **really** slows it down. `10_000` rows took 19:10.06 to complete!
> 
> Issues should be fixed.

I'll test them, although it won't be until tomorrow morning now.

**Comment by ludwiglierhammer** [comment]
No pressure. I'll give it a try myself.

**Comment by ludwiglierhammer** [comment]
> > > I'm happy to discuss polars at some point if you want.
> > 
> > 
> > I tried to implement polars yesterday withour any success 😞. As you said it is quite different to pandas and the implementation needs some more effort. Maybe we should discuss how to imporve the performance in general. What about end of next week?
> 
> End of the next week is good for me. Usually I'm the only person in my office on a Friday so that would be most convenient. Drop me an email with some times that work for you.

There are craftsmen in my apartment today. Unfortunately, it's very noisy. Can we meet next week? Sorry for the inconvenience.

**Comment by jtsiddons** [comment]
> There are craftsmen in my apartment today. Unfortunately, it's very noisy. Can we meet next week? Sorry for the inconvenience.

That's fine - let me know what works for you - mornings are generally better for me.

**Comment by ludwiglierhammer** [comment]
In #348, we try to replace ``TextFileReader`` with ``ParquetStreamReader`` becuase we want to keep the chunking option.
